prow/crier/controller_test.go

```diff
@@ -0,0 +1,244 @@
+	"sync"
+	"testing"
+	"time"
+)
+
+const (
+	testTimeout    = time.Second
+	closeTimeLimit = time.Second
+	controllerName = "CrierTest"
+)
+
+// Fake Reporter
+// Sets: Logic; Some series of test jobs should be reported, while others shouldn't be
+// Asserts: The correct jobs have Report called on them
+type triggerReporter struct {
+	reportedItem *bool
+}
+
+func (f triggerReporter) Report(pj *prowv1.ProwJob) ([]*prowv1.ProwJob, error) {
+	if f.reportedItem != nil {
+		*f.reportedItem = true
+	}
+	return nil, nil
+}
+
+func (f triggerReporter) GetName() string {
+	return "triggerReporter"
+}
+
+func (f triggerReporter) ShouldReport(pj *prowv1.ProwJob) bool {
+	return pj.Status.State == prowv1.TriggeredState
+}
+
+// Fake Informer
+// Sets: The Prow Job Test Cases
+// Asserts: The Controller registers for the queue to be populated
+type fakeInformer struct {
+	eventChannel chan string
+}
+
+func (f fakeInformer) Get(name string) (*prowv1.ProwJob, error) {
+	jobs := map[string]*prowv1.ProwJob{
+		"TestJob1": {
+			Spec: prowv1.ProwJobSpec{
+				Job: "TestJob1",
+			},
+			Status: prowv1.ProwJobStatus{
+				State: prowv1.FailureState,
+			},
+		},
+		"TestJob2": {
+			Spec: prowv1.ProwJobSpec{
+				Job: "TestJob2",
+			},
+			Status: prowv1.ProwJobStatus{
+				State: prowv1.TriggeredState,
+			},
+		},
+	}
+
+	pj, found := jobs[name]
+	if !found {
+		return nil, errors.New("Can't Find ProwJob")
+	}
+	return pj, nil
+}
+
+func (f fakeInformer) ProwJobs(namespace string) prowLister.ProwJobNamespaceLister {
+	return f
+}
+
+func (f fakeInformer) Informer() cache.SharedIndexInformer {
+	return f
+}
+
+func (f fakeInformer) Lister() prowLister.ProwJobLister {
+	return f
+}
+
+func (f fakeInformer) AddEventHandler(handler cache.ResourceEventHandler) {
+	if f.eventChannel != nil {
+		f.eventChannel <- "Registered"
+	}
+}
+
+func (f fakeInformer) Run(stopCh <-chan struct{}) {
+	if f.eventChannel != nil {
+		f.eventChannel <- "Run"
+	}
+}
+
+func (f fakeInformer) AddEventHandlerWithResyncPeriod(handler cache.ResourceEventHandler, resyncPeriod time.Duration) {
+	if f.eventChannel != nil {
+		f.eventChannel <- "Registered"
+	}
+}
+
+func (fakeInformer) HasSynced() bool {
+	return true
+}
+
+func (fakeInformer) GetStore() cache.Store                                            { return nil }
+func (fakeInformer) GetController() cache.Controller                                  { return nil }
+func (fakeInformer) LastSyncResourceVersion() string                                  { return "" }
+func (fakeInformer) AddIndexers(indexers cache.Indexers) error                        { return nil }
+func (fakeInformer) GetIndexer() cache.Indexer                                        { return nil }
+func (fakeInformer) List(selector labels.Selector) (ret []*prowv1.ProwJob, err error) { return nil, nil }
+
+func TestController_processNextItem(t *testing.T) {
+	tests := []struct {
+		name         string
+		expectReport bool
+	}{
+		{
+			name:         "TestJob1",
```

Comment:
Ideally I can read this item in this table and determine what its trying to test and why. Thus IMO it would be better to shove the prowjob into this, not a name that references a magical job defined elsewhere

Reply:
I agree that "TestJob1" and "TestJob2" are bad job names. Thank you

I've also moved the table declarations so that they're all closer to each other, but I strongly believe that keeping the tables separate cuts down on a lot of redundancy and makes it overall more understandable. Again, please give me your feedback on its readability.

Reply:
I disagree that cutting down on this redundancy is a goal of testing. We should minimize the time required to read and modify this test, which is generally easier with some repetition.